lib/tbot/tbot.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tbot
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/tbot/config"
+	"github.com/gravitational/teleport/lib/tbot/identity"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type Bot struct {
+	cfg        *config.BotConfig
+	log        logrus.FieldLogger
+	reloadChan chan struct{}
+
+	// These are protected by getter/setters with mutex locks
+	mu      sync.Mutex
+	_client auth.ClientI
+	_ident  *identity.Identity
+	started bool
+}
+
+func New(cfg *config.BotConfig, log logrus.FieldLogger, reloadChan chan struct{}) *Bot {
+	if log == nil {
+		log = utils.NewLogger()
+	}
+
+	return &Bot{
+		cfg:        cfg,
+		log:        log,
+		reloadChan: reloadChan,
+	}
+}
+
+func (b *Bot) client() auth.ClientI {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b._client
+}
+
+func (b *Bot) setClient(client auth.ClientI) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b._client = client
+}
+
+func (b *Bot) ident() *identity.Identity {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b._ident
+}
+
+func (b *Bot) setIdent(ident *identity.Identity) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b._ident = ident
+}
+
+func (b *Bot) markStarted() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.started {
+		return trace.Errorf("bot has already been started")
```

Comment:
I won't pretend to understand the full rationale, but I'm told we should prefer `trace.BadParameter()` if no other error type fits:
```suggestion
		return trace.BadParameter("bot has already been started")
```

Reply:
Hmm, if that's a pattern we are using then that makes sense. In the gRPC world, I'd probably call this "INVALID_PRECONDITION" rather than "INVALID_ARGUMENT", which is the closest thing to the BadParameter we have.